### ISSUE TITLE
deno vscode installation and downloads behind a coporate proxy

### ISSUE THREAD

**Issue by kraabrsg** [denoland/deno#26935]
Version: Deno x2.0.5

Hi,

sorry if i missed an issue....
We have a node project where we build several js files to run on a RHEL Linux.

The idea is to use deno to build native executeable files via crosscompile.

Now it seems as if deno could not do anything which needs to fetch files e.g.` deno install` and `deno info` wants do download
 from registry.npmjs.org but stucks without doing anything.

I had a phone call with colleague responsible for proxy configuration and told me for which certificate i need to use on my pc to bypass issues like "local certificate errors" .

so i added the .crt file with a deno setting deno.tlsCertificate in the project and also for node via cafile=xxx in npm settings file.

Denos output:
```
tarting Deno language server...
  version: 2.0.5 (release, x86_64-pc-windows-msvc)
  executable: C:\mypath\deno\deno.EXE
Connected to "Visual Studio Code" 1.94.2
Enabling import suggestions for: https://deno.land
  Error fetching registry config for "https://deno.land": EOF while parsing a value at line 1 column 0
Refreshing configuration tree...
  Resolved Deno configuration file: "file:///C:/mypath1/deno.json"
  Resolved package.json: "file:///C:/mypath1/package.json"
  Resolved .npmrc: "C:\Users\myuser\.npmrc"
Could not set npm package requirements: Error getting response at http://registry.npmjs.org/axios for package "axios": An npm specifier not found in cache: "axios", --cached-only is specified.
Could not set npm package requirements: Error getting response at http://registry.npmjs.org/axios for package "axios": An npm specifier not found in cache: "axios", --cached-only is specified.
Server ready.
.....
Download http://registry.npmjs.org/pg-promise
successfully cancelled request with ID: 652
Download http://registry.npmjs.org/@types%2fnode
Download http://registry.npmjs.org/@types%2fpg
Download http://registry.npmjs.org/@types%2fpg-promise
Download http://registry.npmjs.org/@typescript-eslint%2feslint-plugin
Download http://registry.npmjs.org/@typescript-eslint%2fparser
Download http://registry.npmjs.org/axios
Download http://registry.npmjs.org/ebcdic-ascii
Download http://registry.npmjs.org/basic-ftp
Download http://registry.npmjs.org/esbuild
Download http://registry.npmjs.org/fast-xml-parser
Download http://registry.npmjs.org/esbuild-node-externals
Download http://registry.npmjs.org/ftp-ts
Download http://registry.npmjs.org/memorystream
Download http://registry.npmjs.org/node-sql-parser
Download http://registry.npmjs.org/pg
Download http://registry.npmjs.org/ts-node
Download http://registry.npmjs.org/pg-promise
...
Could not set npm package requirements: Error getting response at http://registry.npmjs.org/@types%2fnode for package "@types/node": An npm specifier not found in cache: "@types/node", --cached-only is specified.
Could not set npm package requirements: Error getting response at http://registry.npmjs.org/@types%2fnode for package "@types/node": An npm specifier not found in cache: "@types/node", --cached-only is specified.
Error caching: Error getting response at http://registry.npmjs.org/@types%2fnode for package "@types/node": error sending request for url (https://registry.npmjs.org/@types%2Fnode): client error (Connect): tcp connect error: Ein Verbindungsversuch ist fehlgeschlagen, da die Gegenstelle nach einer bestimmten Zeitspanne nicht richtig reagiert hat, oder die hergestellte Verbindung war fehlerhaft, da der verbundene Host nicht reagiert hat. (os error 10060): Ein Verbindungsversuch ist fehlgeschlagen, da die Gegenstelle nach einer bestimmten Zeitspanne nicht richtig reagiert hat, oder die hergestellte Verbindung war fehlerhaft, da der verbundene Host nicht reagiert hat. (os error 10060)


```
Terminal output stuck: 

```
Download ⣟ [10:36] 0/18
 - http://registry.npmjs.org/@types%2fnode
 - http://registry.npmjs.org/@types%2fpg
 - http://registry.npmjs.org/@types%2fpg-promise
 - http://registry.npmjs.org/@typescript-eslint%2feslint-plugin
```

package.json
```
	"devDependencies": {
		"@types/node": "^20.11.20",
		"@types/pg": "^8.11.10",
		"@types/pg-promise": "^5.4.3",
               ...
      },
	"dependencies": {
               ...
		"pg": "^8.13.1",
		"pg-promise": "^11.10.1"
	}
```

So i am not sure if there is a problem with proxy(is there a checklist or something like that somewhere) , or a misconfiguration in the package.json ?

Thanks!

**Comment by kraabrsg** [comment]
So after further research the os error 10060 is because it is blocked by the coporate firewall ( a proxy is not used here, it is not really necessery for the first steps). 

The stuck download is because the coporare proxy gets a 301, is there a way to avoid that ?

Thanks!

**Comment by bartlomieju** [comment]
@nayeemrmn PTAL, I think we might not be respecting some settings here?

**Comment by nayeemrmn** [comment]
@kraabrsg Use `"deno.env": { "HTTP_PROXY": "http://myproxy:8080" }` instead for now. I don't remember what the situation was for `http.proxy`, if and why we don't support it..

**Comment by kraabrsg** [comment]
Hi @bartlomieju , @nayeemrmn  thanks for getting back on this, but this does not seem to make a difference.

![image](https://github.com/user-attachments/assets/e82643c7-65b7-43d4-81c5-e59740265502)

I tried `deno install `  

![image](https://github.com/user-attachments/assets/7bad7736-b1fd-407c-81c2-a1c46bfbb3a2)

Deno Version: 2.1.0


One big Problem (even if this solution would work) is the ip above which is on a black list 
https://www.virustotal.com/gui/ip-address/34.149.100.209


I have to ask again with our firewall colleaugue to look if something has changed, but seems not... 
Any further ideas to track this down ? 

Btw. deno looks really awesome, it would be a pitty if we could not use this...

**Comment by kraabrsg** [comment]
Hi @nayeemrmn any update on this one ? nothing changed here, it seems as if deno does not reach a proxy at all (according to the colleague). Is there any further option ?

**Comment by nayeemrmn** [comment]
@kraabrsg Try:
```json
{
  "deno.env": {
    "HTTP_PROXY": "http://myproxy:8080",
    "HTTPS_PROXY": "http://myproxy:8080",
  }
}
```

**Comment by kraabrsg** [comment]
@nayeemrmn  thanks , but as far as i know we don't have a https proxxy, we had issues with a node app, because the admins set an environment variable with a https proxy like this.


without https:

```
Starting Deno language server...
  version: 2.1.2 (release, x86_64-pc-windows-msvc)
  executable: C:\mypath\deno\deno.EXE
Connected to "Visual Studio Code" 1.94.2
Enabling import suggestions for: https://deno.land
  Error fetching registry config for "https://deno.land": EOF while parsing a value at line 1 column 0
Refreshing configuration tree...
  Resolved Deno configuration file: "file:///C:/mypath2/vmd-server-deno/deno.json"
  Resolved .npmrc: "C:\Users\myuser\.npmrc"
Could not set npm package requirements: Error getting response at https://registry.npmjs.org/@types%2fnode for package "@types/node": An npm specifier not found in cache: "@types/node", --cached-only is specified.
Server ready.
Failed to check for upgrades: error sending request for url (https://dl.deno.land/release-latest.txt?lsp): client error (Connect): tcp connect error: Ein Verbindungsversuch ist fehlgeschlagen, da die Gegenstelle nach einer bestimmten Zeitspanne nicht richtig reagiert hat, oder die hergestellte Verbindung war fehlerhaft, da der verbundene Host nicht reagiert hat. (os error 10060): Ein Verbindungsversuch ist fehlgeschlagen, da die Gegenstelle nach einer bestimmten Zeitspanne nicht richtig reagiert hat, oder die hergestellte Verbindung war fehlerhaft, da der verbundene Host nicht reagiert hat. (os error 10060)

```

with https
```
Starting Deno language server...
  version: 2.1.2 (release, x86_64-pc-windows-msvc)
  executable: C:\mypath\deno\deno.EXE
Connected to "Visual Studio Code" 1.94.2
Enabling import suggestions for: https://deno.land
  Error fetching registry config for "https://deno.land": EOF while parsing a value at line 1 column 0
Refreshing configuration tree...
  Resolved Deno configuration file: "file:///C:/mypath2/vmd-server-deno/deno.json"
  Resolved .npmrc: "C:\Users\myuser\.npmrc"
Could not set npm package requirements: Error getting response at https://registry.npmjs.org/@types%2fnode for package "@types/node": An npm specifier not found in cache: "@types/node", --cached-only is specified.
Server ready.
Failed to check for upgrades: error sending request for url (https://dl.deno.land/release-latest.txt?lsp): client error (Connect): invalid peer certificate: UnknownIssuer

```

would it be worth investigating this with wireshark  ?

**Comment by nayeemrmn** [comment]
@kraabrsg Try moving the other options to their equivalent env vars:
```json
{
  "deno.env": {
    "HTTP_PROXY": "http://myproxy:8080",
    "HTTPS_PROXY": "http://myproxy:8080",
    "DENO_CERT": "<...value of deno.tlsCertificate>",
    "DENO_TLS_CA_STORE": "system"
  }
}
```

**Comment by kraabrsg** [comment]
@nayeemrmn thanks but does not solve ist

with 
```
    "deno.env": {
        "HTTP_PROXY": "http://myproxy.at:8080",
        "DENO_TLS_CA_STORE": "system",
        "DENO_CERT": "C:\\mypath\\CertThatisCheckedOnProxy.crt"
    }
```

yields to exactly the same as above (without https)

**Comment by nayeemrmn** [comment]
@kraabrsg Does this not work in CLI as well? Like with `deno check` or `deno install`.

**Comment by kraabrsg** [comment]
@nayeemrmn 
no it does not work with  `deno check ` (it takes a while and tries to fetch jsr:@std/path/meta.json and jsr:@std/fs/meta.json):

```
PS C:\Emypath\vmd-server-deno> deno check .\src\UserPassStore.ts
error: JSR package manifest for '@std/fs' failed to load. Import 'https://jsr.io/@std/fs/meta.json' failed: error sending request for url (https://jsr.io/@std/fs/meta.json): client error (Connect): tcp connect error: Ein Verbindungsversuch ist fehlgeschlagen, da die Gegenstelle nach einer bestimmten Zeitspanne nicht richtig reagiert hat, oder die hergestellte Verbindung war fehlerhaft, da der verbundene Host nicht reagiert hat. (os error 10060): Ein Verbindungsversuch ist fehlgeschlagen, da die Gegenstelle nach einer bestimmten Zeitspanne nicht richtig reagiert hat, oder die hergestellte Verbindung war fehlerhaft, da der verbundene Host nicht reagiert hat. (os error 10060)
    at file:///C:/mypath/vmd-server-deno/src/UserPassStore.ts:1:21

```


and `deno install` also takes a while but does nothing as it seems

![image](https://github.com/user-attachments/assets/9dd4e0aa-106f-459b-91c1-4ff74cb5efaa)

**Comment by bartlomieju** [comment]
Closing as stale, the thread has had no activity since December 2024.

Deno does support proxy and TLS-certificate configuration through environment variables, documented here: https://docs.deno.com/runtime/reference/env_variables/

The relevant ones for this setup are:

- `HTTP_PROXY` / `HTTPS_PROXY` - proxy address for HTTP/HTTPS requests (module downloads and `fetch`)
- `NO_PROXY` - comma-separated list of hosts that bypass the proxy
- `DENO_CERT` - load a certificate authority from a PEM-encoded file
- `DENO_TLS_CA_STORE` - certificate stores to use (e.g. `system`)

In the LSP these can be set via `deno.env` in your VS Code settings, and they apply to the CLI directly as environment variables.

Based on the discussion, the remaining blocker here was environmental rather than a Deno configuration problem: your firewall was blocking the destination IP (`34.149.100.209`, a load-balancer address that jsr.io / deno.land / dl.deno.land resolve behind) with an `os error 10060` connect timeout. That can only be resolved by having the address allow-listed on your network, no Deno setting can work around a blocked TCP connection.

If proxy or certificate configuration still doesn't work for you on a current Deno (with the env vars above and the destination reachable), please open a new issue with the exact config you used and the output of a CLI command like `deno install` or `deno check`, and we'll investigate. Thanks!
